include/zephyr/spinlock.h

```diff
@@ -183,7 +192,15 @@ static ALWAYS_INLINE void k_spin_unlock(struct k_spinlock *l,
 	ARG_UNUSED(l);
 #ifdef CONFIG_SPIN_VALIDATE
 	__ASSERT(z_spin_unlock_valid(l), "Not my spinlock %p", l);
-#endif
+
+#if defined(CONFIG_SPIN_LOCK_TIME_LIMIT) && (CONFIG_SPIN_LOCK_TIME_LIMIT != 0)
+	uint32_t delta = sys_clock_cycle_get_32() - l->lock_time;
```

Comment:
Hmm, is it OK that we don't care about rollover here? `***cycle_get_32` represents the fastest system timer and on platforms with multi GHz clock rollover happens often.

Reply:
If a spin lock is held for more than one roll over of the timer clock then I think you've probably held a spin lock for too long already. This would be, at 4GHz, about a full second.

Reply:
Ok, just imagine the situation:

* we lock the spinlock. The sys_clock_cycle_get_32 returns `0xFFFF'FFF0`, so we write this value to `l->lock_time`
* we unlock spinlock after 50 cycles. The sys_clock_cycle_get_32 returns `0x0000'0022`, so `delta = 0x0000'0022 - 0xFFFF'FFF0`.

As I can see we'll get false positive here.

Reply:
delta will return 50 with the above math given unsigned integer math according to the C specs here. The only time this is an issue is if the count wraps multiple times, which won't be seen. But that's unlikely even with the fastest clock speeds today to happen in a time frame that spin locks should be held for.

Reply:
The standard actually says this is how unsigned math for C is supposed to work, so we're good! See 6.2.5 paragraph 9 for the C99 ISO spec